### PR TITLE
Editor.js as editor

### LINKED ISSUE #2945: Switch to block style editor
## Feature Request

| Q           | A      |
| ----------- | ------ |
| New Feature | yes    |
| RFC         | no |
| BC Break    | yes |

### Summary

CK Editor is kinda bloated and a huge block of code we can't properly test. By switching to SirTrevor everything will be more modular

### PR DESCRIPTION
**Pull request by carakas**
## Type

- Feature

## Resolves the following issues

fixes #2945 

## Pull request description

Basic implementation of a block editor with paragraphs, headers, media library image and lists

https://editorjs.io/

### CODE REVIEW

**Comment by jessedobbelaere** [comment]
@carakas I'm wondering, did you go for editorjs or sir trevor now? Because the diff seems to mention both in the code (package.json vs gulp tasks etc)? 👀

**Comment by carakas** [comment]
@jessedobbelaere sorry, I started off with sir-trevor but then realised it's kinda abandoned, the implemented editor there is also deprecated so I switched. Must have forgotten some code, fixed in ee7b24dcc

File: src/Backend/Core/Js/editor/editor.js [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    return scope[scopeSplit[scopeSplit.length - 1]]
+  }
+
+  static fromJson ($element, jsonConfig) {
+    let config = JSON.parse(jsonConfig)
+    for (const name of Object.keys(config)) {
+      config[name].class = this.getClassFromVariableName(config[name].class)
+    }
+
+    this.create($element, config)
```
[the remark this answers is not in the document]
Comment (carakas) [reply src/Backend/Core/Js/editor/editor.js]: Yes

File: src/Common/BlockEditor/Twig/ParseBlocksExtension.php [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+            return $json;
+        }
+
+        /** @var AdapterInterface $cache */
+        $cache = $this->container->get('cache.block_editor');
+        $cacheKey = md5($json);
+        $cachedBlock = $cache->getItem($cacheKey);
+        if ($cachedBlock->isHit()) {
+            $parsedData = $cachedBlock->get();
+            if ($parsedData['time'] === $data['time'] && $parsedData['version'] === $data['version']) {
```
[the remark this answers is not in the document]
Comment (carakas) [reply src/Common/BlockEditor/Twig/ParseBlocksExtension.php]: If we have a hit and something else is in there somebody has been tampering with the cache, I don't mind if it crashes in that case

File: src/Backend/Core/Js/backend.js [hunk 3]
```diff
@@ -565,6 +569,34 @@ jsBackend.ckeditor = {
...
+
+    jsBackend.blockEditor.loadEditorsInCollections()
+  },
+
+  createEditor: function ($element) {
+    BlockEditor.editor.fromJson($element, $element.attr('fork-block-editor-config'))
+  },
+
+  loadEditorsInCollections: function () {
+    $('[data-addfield=collection]').on('collection-field-added', function () {
```
[the remark this answers is not in the document]
Comment (carakas) [reply src/Backend/Core/Js/backend.js]: fixed in fda209032

File: src/Common/BlockEditor/Blocks/AbstractBlock.php [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+    /** The url to the JS file with the config needed to make this block work in the editor */
+    abstract public function getJavaScriptUrl(): ?string;
+
+    abstract public function parse(array $data): string;
+
+    final protected function parseWithTwig(string $template, array $data): string
+    {
+        /** @var TwigTemplate $templating */
+        $templating = $this->container->get('templating');
```
[the remark this answers is not in the document]
Comment (carakas) [reply src/Common/BlockEditor/Blocks/AbstractBlock.php]: Fixed in a07e53fe5
I think initially something didn't work but with the current implementation it works

**Pull request merged**